[P037] Fix event generation bug, add event-limit options

tonhuisman opened the pull request:
Resolves #3999 

- Fix bug that events where always generated, even when option was disabled
- Add optional De-duplication of events
- Add optional Event-queue-depth limit
- Fix (change) that events `MQTTImport#Connected` and `MQTTImport#Disconnected` are only generated if the plugin-task is enabled
- Enable all compile-time optional features in MAX builds
- Add optional replacement of a character in events by a comma

TODO:
- [x] Add option to replace event-value character by a comma (to allow further processing as arguments by other events)
- [x] Adjust configuration so all options are included in MAX builds
- [x] Update documentation
- [ ] Testing (feedback)

tonhuisman committed:
[P037] Fix event generation bug, add event-limit options

tonhuisman committed:
[P037] Only generate event MQTTImport#(dis)connected if plugin-task is enabled

tonhuisman committed:
[P037] Enable all optional features in MAX builds

tonhuisman committed:
[P037] Fix Linux compilation issue (tighter type-checks)

TD-er commented on src/_P037_MQTTImport.ino:
```diff
@@ -53,6 +59,75 @@ String P037_getMQTTLastTopicPart(const String& topic) {
+      }
+      # endif // if !defined(LIMIT_BUILD_SIZE) || defined(P037_OVERRIDE)
+
+      result = !queuedEvent.equalsIgnoreCase(newEvent); // IgnoreCase-compare!
```
Hmm too bad you can't use `find` here on the queue, simply because you need to match case insensitive.
But I guess it is good to have such a dedup-option for events anyway, so why not add this dedup option to the add function of the eventQueue? (default parameter for dedup set to false, but we may later make it a setting)

tonhuisman replied:
Hm, maybe I can use `find`, as the events will always be the same, because they come from an external source, and often device-generated.

tonhuisman replied:
Switched to case-sensitive comparison. Now use `find` for matching events.

TD-er commented on src/_P037_MQTTImport.ino:
```diff
@@ -53,6 +59,75 @@ String P037_getMQTTLastTopicPart(const String& topic) {
+  }
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
+    String log = F("MQTT: Event added: ");
```
Please wrap these logs in `#ifndef BUILD_NO_DEBUG`

tonhuisman replied:
I wanted this one to always be available, that's why I left that #define out.

TD-er commented on src/src/DataStructs/EventQueue.h:
```diff
@@ -23,6 +23,18 @@ struct EventQueueStruct {
+
+  std::list<String>::iterator begin() {
+    return _eventQueue.begin();
+  }
```
These can be a `const_iterator begin() const` as you're not doing anything with the iterator, nor deleting it from the queue.

tonhuisman replied:
Removed, as no longer needed

TD-er commented on src/src/DataStructs/EventQueue.h:
```diff
@@ -23,6 +23,18 @@ struct EventQueueStruct {
+    return _eventQueue.begin();
+  }
+
+  std::list<String>::iterator end() {
```
ditto

tonhuisman committed:
[P037] Add option to replace value (special) characters by a comma

tonhuisman committed:
[P037] Add option to replace value (special) characters by a comma

tonhuisman committed:
[P037] Add option to replace value (special) characters by a comma

tonhuisman committed:
[EventQueue] Optionally deduplicate event against current event list (case-sensitive)

tonhuisman committed:
[P037] Use EventQueue deduplicate feature, case-sensitive

TD-er commented on src/_P037_MQTTImport.ino:
```diff
@@ -61,52 +62,22 @@ String P037_getMQTTLastTopicPart(const String& topic) {
+
+      if (P037_REPLACE_BY_COMMA != 0x0) {
+        String character = F(" ");
+        character[0] = static_cast<uint8_t>(P037_REPLACE_BY_COMMA);
```
`const String character = static_cast<char>(P037_REPLACE_BY_COMMA);`

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into bugfix/p037-event-generation-overloads-esp

tonhuisman committed:
[P037] Code optimizations, allow a few extra replace-characters

tonhuisman committed:
[P037] Small UI adjustments

tonhuisman committed:
[P037] Update documentation

tonhuisman committed:
[P037] Improvements of documentation

tonhuisman committed:
Merge branch 'mega' into bugfix/p037-event-generation-overloads-esp

TD-er commented on src/_P037_MQTTImport.ino:
```diff
@@ -531,7 +609,7 @@ boolean Plugin_037(uint8_t function, struct EventStruct *event, String& string)
                   RuleEvent += '=';
                   RuleEvent += unparsedPayload;
-                  eventQueue.addMove(std::move(RuleEvent));
+                  P037_addEventToQueue(event, RuleEvent);
```
I think these kind of events need the eventvalues wrapped in quotes if they contain a comma or space.
Also they need to be wrapped in quotes not present in the payload and only if not already wrapped with quotes (there is a function for it)

TD-er committed:
Wrap eventvalues in quotes when needed for generated events

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into bugfix/p037-event-generation-overloads-esp

tonhuisman committed:
[Helpers] Enable unconditional quote-wrapping of a String value

tonhuisman committed:
[Helpers] Refactor quote-wrapping of a String value

tonhuisman committed:
Merge branch 'mega' into bugfix/p037-event-generation-overloads-esp

TD-er committed:
[Cleanup] Remove duplicate code to wrap text in quotes

TD-er committed:
[MQT import] Reduce memory usage, especially during load/save

TD-er committed:
[Cleanup] Fix tricky PLUGIN_EXIT calls

Plugin task data will be destructed after calling PLUGIN_EXIT.
If it was destructed before that and not properly set to a nullptr, then you might get a crash when saving settings.

TD-er committed:
[MQTT Import] Unsubscribe from MQTT topics on exit

TD-er committed:
[MQTT import] Add log when task cannot unsubscribe as others use it
